ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java

```diff
@@ -239,12 +248,243 @@ public void shouldBuildCorrectFormatsForV2NonWindowedTable() {
+        contextStacker
+    );
+
+    // Then:
+    assertThat(((WindowedTableSource) result.getSourceStep()).getPseudoColumnVersion(), equalTo(OLD_PSEUDOCOLUMN_VERSION));
+    assertValidSchema(result);
+  }
+
+  @Test
+  public void shouldCreateWindowedTableSourceWithNewPseudoColumnVersionIfNoOldQuery() {
+    // Given:
+    givenWindowedTable();
+    givenOnNewPseudoColumnVersion();
+
+    // When:
+    final SchemaKStream<?> result = SchemaKSourceFactory.buildSource(
+        buildContext,
+        dataSource,
+        contextStacker
+    );
+
+    // Then:
+    assertThat(((WindowedTableSource) result.getSourceStep()).getPseudoColumnVersion(), equalTo(NEW_PSEUDOCOLUMN_VERSION));
+    assertValidSchema(result);
+  }
+
+
+
+  @Test
+  public void shouldThrowOnV1TableSourceWithPseudoColumnVersionGreaterThanZero() {
+    // Given:
+    givenNonWindowedTable();
+    givenExistingTableSourceV1();
+    when(tableSourceV1.getPseudoColumnVersion()).thenReturn(NEW_PSEUDOCOLUMN_VERSION);
+
+    // When:
+    final Exception e = assertThrows(
+        IllegalStateException.class,
+        () -> SchemaKSourceFactory.buildSource(
+            buildContext,
+            dataSource,
+            contextStacker
+        )
+    );
+
+    // Then:
+    assertThat(
+        e.getMessage(),
+        containsString("TableSourceV2 was released in conjunction with pseudocolumnversion 1. Something has gone very wrong")
+    );
+  }
+
+  private void givenNonWindowedStream() {
+    when(dataSource.getDataSourceType()).thenReturn(DataSourceType.KSTREAM);
+    when(keyFormat.getWindowInfo()).thenReturn(Optional.empty());
+    when(keyFormat.isWindowed()).thenReturn(false);
+  }
+
+  private void givenWindowedStream() {
+    when(dataSource.getDataSourceType()).thenReturn(DataSourceType.KSTREAM);
+    when(keyFormat.isWindowed()).thenReturn(true);
+    when(keyFormat.getWindowInfo()).thenReturn(Optional.of(windowInfo));
+  }
+
   private void givenNonWindowedTable() {
     when(dataSource.getDataSourceType()).thenReturn(DataSourceType.KTABLE);
     when(keyFormat.getFormatInfo().getFormat()).thenReturn("JSON");
     when(keyFormat.isWindowed()).thenReturn(false);
   }
 
+  private void givenWindowedTable() {
+    when(dataSource.getDataSourceType()).thenReturn(DataSourceType.KTABLE);
+    when(keyFormat.isWindowed()).thenReturn(true);
+    when(keyFormat.getWindowInfo()).thenReturn(Optional.of(windowInfo));
+  }
+
+  private void givenExistingTableSourceV1() {
+    givenExistingQueryWithOldPseudoColumnVersion();
+    when(buildContext.getPlanInfo()).thenReturn(Optional.of(planInfo));
+    when(planInfo.getSourceSet()).thenReturn(ImmutableSet.of(tableSourceV1));
+    when(tableSource.getPseudoColumnVersion()).thenReturn(OLD_PSEUDOCOLUMN_VERSION);
+  }
+
+  private void givenExistingQueryWithOldPseudoColumnVersion() {
+    when(buildContext.getPlanInfo()).thenReturn(Optional.of(planInfo));
+    when(planInfo.getSourceSet()).thenReturn(ImmutableSet.of(streamSource));
```

Comment:
Why is this method always using `streamSource` even from the tests that are testing tables, etc? I think we should make this wiring more airtight -- you can pass in the source step to mock as a param if you don't want to make copies of this method.